src/hmem_cuda_gdrcopy.c

```diff
@@ -0,0 +1,292 @@
+	if (!gdrapi_handle) {
+		FI_INFO(&core_prov, FI_LOG_CORE,
+			"Failed to dlopen libgdrapi.so\n");
+		return -FI_ENOSYS;
+	}
+
+	gdrcopy_ops.gdr_open = dlsym(gdrapi_handle, "gdr_open");
+	if (!gdrcopy_ops.gdr_open) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "Failed to find gdr_open\n");
+		goto err_dlclose_gdrapi;
+	}
+
+	gdrcopy_ops.gdr_close = dlsym(gdrapi_handle, "gdr_close");
+	if (!gdrcopy_ops.gdr_close) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "Failed to find gdr_close\n");
+		goto err_dlclose_gdrapi;
+	}
+
+	gdrcopy_ops.gdr_pin_buffer = dlsym(gdrapi_handle, "gdr_pin_buffer");
+	if (!gdrcopy_ops.gdr_pin_buffer) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "Failed to find gdr_pin_buffer\n");
+		goto err_dlclose_gdrapi;
+	}
+
+	gdrcopy_ops.gdr_unpin_buffer = dlsym(gdrapi_handle, "gdr_unpin_buffer");
+	if (!gdrcopy_ops.gdr_unpin_buffer) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "Failed to find gdr_unpin_buffer\n");
+		goto err_dlclose_gdrapi;
+	}
+
+	gdrcopy_ops.gdr_map = dlsym(gdrapi_handle, "gdr_map");
+	if (!gdrcopy_ops.gdr_map) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "Failed to find gdr_map\n");
+		goto err_dlclose_gdrapi;
+	}
+
+	gdrcopy_ops.gdr_unmap = dlsym(gdrapi_handle, "gdr_unmap");
+	if (!gdrcopy_ops.gdr_unmap) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "Failed to find gdr_unmap\n");
+		goto err_dlclose_gdrapi;
+	}
+
+	gdrcopy_ops.gdr_copy_to_mapping = dlsym(gdrapi_handle, "gdr_copy_to_mapping");
+	if (!gdrcopy_ops.gdr_copy_to_mapping) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "Failed to find gdr_copy_to_mapping\n");
+		goto err_dlclose_gdrapi;
+	}
+
+	gdrcopy_ops.gdr_copy_from_mapping = dlsym(gdrapi_handle, "gdr_copy_from_mapping");
+	if (!gdrcopy_ops.gdr_copy_from_mapping) {
+		FI_WARN(&core_prov, FI_LOG_CORE, "Failed to find gdr_copy_from_mapping\n");
+		goto err_dlclose_gdrapi;
+	}
+
+	return FI_SUCCESS;
+
+err_dlclose_gdrapi:
+	memset(&gdrcopy_ops, 0, sizeof(gdrcopy_ops));
+	dlclose(gdrapi_handle);
+	return -FI_ENODATA;
+#else /* ENABLE_GDRCOPY_DLOPEN */
+	return FI_SUCCESS;
+#endif /* ENABLE_GDRCOPY_DLOPEN */
+}
+
+int cuda_gdrcopy_hmem_cleanup(void)
+{
+#ifdef ENABLE_GDRCOPY_DLOPEN
+	dlclose(gdrapi_handle);
+#endif
+	return FI_SUCCESS;
+}
+
+int cuda_gdrcopy_to_dev(uint64_t devhandle, void *devptr, const void *hostptr, size_t len)
+{
+	ssize_t off;
+	struct ofi_gdrcopy_handle *gdrcopy;
+	void *gdrcopy_user_ptr;
+
+	if (!gdrcopy_ops.gdr_copy_to_mapping)
+		return -FI_ENOSYS;
+
+	gdrcopy = (struct ofi_gdrcopy_handle *)devhandle;
+	off = (char *)devptr - (char *)gdrcopy->cuda_ptr;
+	assert(off >= 0 && off + len <= gdrcopy->length);
+	gdrcopy_user_ptr = (char *)gdrcopy->user_ptr + off;
+	gdrcopy_ops.gdr_copy_to_mapping(gdrcopy->mh, gdrcopy_user_ptr, hostptr, len);
+	return 0;
+}
+
+int cuda_gdrcopy_from_dev(uint64_t devhandle, void *hostptr, const void *devptr, size_t len)
+{
+	ssize_t off;
+	struct ofi_gdrcopy_handle *gdrcopy;
+	void *gdrcopy_user_ptr;
+
+	if (!gdrcopy_ops.gdr_copy_from_mapping)
+		return -FI_ENOSYS;
```

Comment:
Can this be an assert?

Reply:
Yes, it can and should be an assert, fixed in new version.